drgn/cli.py

```diff
@@ -0,0 +1,151 @@
+    except UnicodeEncodeError:
+        encoded = text.encode(sys.stdout.encoding, "backslashreplace")
+        if hasattr(sys.stdout, "buffer"):
+            sys.stdout.buffer.write(encoded)
+        else:
+            text = encoded.decode(sys.stdout.encoding, "strict")
+            sys.stdout.write(text)
+    sys.stdout.write("\n")
+    setattr(builtins, "_", value)
+
+
+def run_cli(
+    prog: drgn.Program,
+    additional_globals: Optional[Dict[str, Any]] = None,
+    banner: Optional[str] = None,
+    additional_banner: Optional[str] = None,
+    quiet: bool = False,
+) -> None:
+    """
+    Run the drgn command line interface.
+
+    This function allows your application to embed a drgn CLI that operates the
+    same as the one which drgn itself provides. Simply provide an already
+    configured :class:`drgn.Program`, and any customizations you would like for
+    the banner and globals, and this function will create REPL and run it until
+    the user exits.
+
+    :param prog: The program to insert into the CLI namespace as ``prog``
+    :param additional_globals: A dictionary to add into the global namespace of
+      the CLI. Drgn will populate the namespace with the normal default: the
+      program, the drgn module, and relevant helpers. The additional globals
+      will be added last, and so they can override any preexisting variable with
+      the same name.
+    :param banner: Replace the normal drgn startup banner with your own. If you
+      simply want to add more information below drgn's banner, see
+      ``additional_banner`` below.
+    :param additional_banner: Add your own text which comes after drgn's default
+      banner
+    """
+    version = drgn_version_header()
+    sys.path.insert(0, "")
+
+    init_globals: Dict[str, Any] = {"prog": prog}
+    init_globals["drgn"] = drgn
+    drgn_globals = [
+        "NULL",
+        "Object",
+        "cast",
+        "container_of",
+        "execscript",
+        "offsetof",
+        "reinterpret",
+        "sizeof",
+    ]
+    for attr in drgn_globals:
+        init_globals[attr] = getattr(drgn, attr)
+    init_globals["__name__"] = "__main__"
+    init_globals["__doc__"] = None
+
+    histfile = os.path.expanduser("~/.drgn_history")
+    try:
+        readline.read_history_file(histfile)
+    except OSError as e:
+        if not isinstance(e, FileNotFoundError) and not quiet:
+            print("could not read history:", str(e), file=sys.stderr)
+
+    def write_history_file() -> None:
+        try:
+            readline.write_history_file(histfile)
+        except OSError as e:
+            if not quiet:
+                print("could not write history:", str(e), file=sys.stderr)
+
+    atexit.register(write_history_file)
+
+    readline.set_history_length(1000)
+    readline.parse_and_bind("tab: complete")
+    readline.set_completer(Completer(init_globals).complete)
+    atexit.register(lambda: readline.set_completer(None))
+
+    sys.displayhook = displayhook
```

Comment:
We modify `sys.path` and `sys.displayhook` globally. Perhaps we should restore them before we return with a `contextlib.ExitStack` handler, as well.

The `readline` settings are another story. I don't see a good way to scope those changes to only affect this function.

I'm imagining a use case where the caller has its own REPL (e.g., @pfactum's [crush](https://codeberg.org/pf-kernel/crush)) and it wants to temporarily drop into the drgn REPL, then return to its own REPL. We can limit a lot of the global state we change, but maybe not `readline`...

Reply:
Yeah, the `readline` stuff can't be entirely be avoided. I think the best bet here is to simply document what drgn clobbers with readline, and recommend that users save & restore their readline settings.

We use the history mechanism, so we should clear history prior to reading our file, and we should clear the history after writing our file. We should unbind the TAB key and set the completer to the system default. I think these are the two major readline features we use. If they're documented, then it's easy for a caller to call their history file persistence routine prior to drgn, and then reload their history file plus completer.

As far as an `ExitStack`... unless we're allowing the user to provide a context manager, I think it may be simplest to just wrap the `code.interact()` with a `try/finally`.